fail2ban/tests/fail2banclienttestcase.py

```diff
@@ -227,54 +246,133 @@ class Fail2banClientServerBase(LogCaptureTestCase):
+
+	#
+	# Common tests
+	#
+	def _testStartForeground(self, tmp, startparams, phase):
+		# start and wait to end (foreground):
+		logSys.debug("start of test worker")
+		phase['start'] = True
+		self.assertExitsNormally(("-f",) + startparams + ("start",))
+		# end :
+		phase['end'] = True
+		logSys.debug("end of test worker")
+
+	@with_tmpdir
+	def testStartForeground(self, tmp):
+		# intended to be ran only in subclasses
+		th = None
+		try:
+			# started directly here, so prevent overwrite test cases logger with "INHERITED"
+			startparams = _start_params(tmp, logtarget="INHERITED")
+			# because foreground block execution - start it in thread:
+			phase = dict()
+			th = Thread(
+				name="_TestCaseWorker",
+				target=self._testStartForeground,
+				args=(tmp, startparams, phase)
+			)
+			th.daemon = True
+			th.start()
+			try:
+				# wait for start thread:
+				Utils.wait_for(lambda: phase.get('start', None) is not None, MAX_WAITTIME)
+				self.assertTrue(phase.get('start', None))
+				# wait for server (socket and ready):
+				self._wait_for_srv(tmp, True, startparams=startparams)
+				self.pruneLog()
+				# several commands to server:
+				self.assertExitsNormally(startparams + ("ping",))
+				self.assertExitsAbnormally(startparams + ("~~unknown~cmd~failed~~",))
+				self.assertExitsNormally(startparams + ("echo", "TEST-ECHO",))
+			finally:
+				self.pruneLog()
+				# stop:
+				self.assertExitsNormally(startparams + ("stop",))
+				# wait for end:
+				Utils.wait_for(lambda: phase.get('end', None) is not None, MAX_WAITTIME)
+				self.assertTrue(phase.get('end', None))
+				self.assertLogged("Shutdown successful", "Exiting Fail2ban")
+		finally:
+			_kill_srv(tmp)
+			if th:
+				th.join()
+
 
 class Fail2banClientTest(Fail2banClientServerBase):
 
+	def assertExitsNormally(self, args):
+		self.assertRaises(ExitException, _exec_client, ((CLIENT,) + args))
+
+	def assertExitsAbnormally(self, args):
+		self.assertRaises(FailExitException, _exec_client, ((CLIENT,) + args))
+
 	def testConsistency(self):
-		self.assertTrue(os.path.isfile(os.path.join(os.path.join(BIN), CLIENT)))
-		self.assertTrue(os.path.isfile(os.path.join(os.path.join(BIN), SERVER)))
+		self.assertTrue(isfile(pjoin(BIN, CLIENT)))
+		self.assertTrue(isfile(pjoin(BIN, SERVER)))
 
 	def testClientUsage(self):
-		self.assertRaises(ExitException, _exec_client, 
-			(CLIENT, "-h",))
+		self.assertExitsNormally(("-h",))
 		self.assertLogged("Usage: " + CLIENT)
 		self.assertLogged("Report bugs to ")
 		self.pruneLog()
-		self.assertRaises(ExitException, _exec_client, 
-			(CLIENT, "-vq", "-V",))
+		self.assertExitsNormally(("-vq", "-V",))
 		self.assertLogged("Fail2Ban v" + fail2bancmdline.version)
 
 	@with_tmpdir
 	def testClientDump(self, tmp):
 		# use here the stock configuration (if possible)
 		startparams = _start_params(tmp, True)
-		self.assertRaises(ExitException, _exec_client, 
-			((CLIENT,) + startparams + ("-vvd",)))
+		self.assertExitsNormally(
```

Comment:
I understood now, it was done also for common testing of foreground execution for both client and server.
Good idea...
Well imho, the names are not so good, so renamed now in execSuccess/execFailed, and little bit rewritten using arguments expanding to increase readability (removed tuple concatenation)